dm_cli/utils/resolve_local_ids.py

```diff
@@ -0,0 +1,136 @@
+    for i, at in enumerate(attribute_types):
+        a = at[0]
+        t = at[1]
+        if is_primitive(t):
+            search_in_primitive(a, pattern, f"{path}[{i}]", references)
+        else:
+            action = type_to_action[t]
+            action(a, pattern, target_attr, f"{path}[{i}]", False, targets, references)
+
+
+def search_in_primitive(val, pattern: str, path: str, references: dict):
+    match = re.match(pattern, str(val))
+    if match:
+        references[path] = match.group(1)
+
+
+def is_primitive(t: type):
+    return t == str or t == int or t == float or t == bool
+
+
+def is_list(t: type):
+    return t == list
+
+
+def is_dict(t: type):
+    return t == dict
+
+
+def search_in_dict(
+    document: dict, pattern: str, target_attr: str, path: str, parent_was_dict: bool, targets: dict, references: dict
+):
+    """
+    Searches for internal references in 'document', referencing to a target attribute.
+
+    [Returns]:
+    Populates two lookup tables:
+    - lookup_of_references
+        mapping from the location of the reference (path), to the local id
+
+    - lookup_of_targets
+        mapping from the local id to the location of the element which owns that id (absolute local path)
+
+    [Inputs]:
+    - document
+        the input document which contains internal references
+
+    - pattern (what does a 'reference' look like?)
+        regex which tries to match the reference structure, defaults to ^.$<some_id>
+        is used to populate the lookup_of_references table
+
+    - target_attr (where is the reference pointing to?)
+        the attribute which contains the id which is being referred to
+        is used to populate the lookup_of_targets table
+
+    - path
+        the absolute path at each level in the search
+
+    - parent_was_dict
+        flag to track the type of the parent
+    """
+
+    attribute_types = map(lambda attr: (attr, type(document[attr])), list(document.keys()))
+    for a, t in attribute_types:
+        val = document[a]
+        if is_primitive(t):
+            if a == target_attr:
+                parent_path = path.rsplit("[", 1)[0]
+                final_path = path if parent_was_dict else f"{parent_path}({target_attr}={val})"
+                targets[val] = final_path
+            search_in_primitive(val, pattern, f"{path}.{a}", references)
+        else:
+            action = type_to_action[t]
+            action(document[a], pattern, target_attr, f"{path}.{a}", True, targets, references)
+    return targets, references
+
+
+def _dig_and_replace(document, path_iterator: iter, replace: str):
+    attr = next(path_iterator, None)
+    if not attr:
+        return True
+    attr = int(attr) if is_list(type(document)) else attr
+    last_attribute = _dig_and_replace(document[attr], path_iterator, replace)
+    if last_attribute:
+        document[attr] = replace
+
+
+def dig_and_replace(targets: dict, references: dict, document):
```

Comment:
Could we add a unit test for `dig_and_replace`? 

Easier to see what's going on then :)